lib/fw.c

```diff
@@ -783,9 +783,92 @@ static int switchtec_fw_info_metadata_gen4(struct switchtec_dev *dev,
 	return -1;
 }
 
+struct switchtec_flash_part_all_info_gen4 {
+	uint32_t firmware_version;
+	uint32_t flash_size;
+	uint8_t ecc_enable;
+	uint8_t rsvd1;
+	uint8_t running_bl2_flag;
+	uint8_t running_cfg_flag;
+	uint8_t running_img_flag;
+	uint8_t rsvd2;
+	uint32_t rsvd3[12];
+	struct switchtec_flash_part_info_gen4  {
+		uint32_t image_crc;
+		uint32_t image_len;
+		uint16_t image_version;
+		uint8_t valid;
+		uint8_t active;
+		uint32_t part_start;
+		uint32_t part_end;
+		uint32_t part_offset;
+		uint32_t part_size_dw;
+		uint8_t read_only;
+		uint8_t is_using;
+		uint8_t rsvd[2];
+	} map0, map1, keyman0, keyman1, bl20, bl21, cfg0, cfg1,
+	  img0, img1, nvlog, vendor[8];
+};
+
 static int switchtec_fw_part_info_gen4(struct switchtec_dev *dev,
 				       struct switchtec_fw_image_info *inf)
 {
+	struct switchtec_flash_part_all_info_gen4 all_info;
+	struct switchtec_flash_part_info_gen4 *part_info;
+
+	int ret;
+	uint8_t subcmd = MRPC_PART_INFO_GET_ALL_INFO;
+
+	ret = switchtec_cmd(dev, MRPC_PART_INFO, &subcmd, sizeof(subcmd),
+			    &all_info, sizeof(all_info));
```

Comment:
Hmm, kind of tricky, but in the future it might be nice to find a way to call this only once when getting all the partitions. Instead of once for every partition and then only using the one partition's info. Just a future thought for a cleanup.

Reply:
I agree.